### ISSUE TITLE
Toolbar inconsistently adds margin to its `div` element on Prism's site

### ISSUE THREAD

**Issue by hoonweiting** [PrismJS/prism#3110]
**Information:**
- Prism version: Latest
- Plugins: Autoloader, Toolbar, Show Language
- Environment: Browser? Prism's site?

**Description**
Toolbar adds some margin to its own `div` element on Prism's site, and it does so in a strange manner. For some code blocks it's `margin: 8px 0px;`, for others it's `margin: 0px;`. I have no idea why. (Screenshots below are from Chrome)

![image](https://user-images.githubusercontent.com/59229084/135910655-0eebda60-4318-44f5-abb8-3d94dac2ed3a.png)

![image](https://user-images.githubusercontent.com/59229084/135912088-fd14a942-c9ee-4b13-8770-4e8c21ba259d.png)

**Example**
You can check out #3104 yourself (though it's a little inconvenient), where I am writing some docs on Prism tokens. This can also be seen on the [Toolbar](https://prismjs.com/plugins/toolbar/), [Show Language](https://prismjs.com/plugins/show-language/), and [Copy to Clipboard](https://prismjs.com/plugins/copy-to-clipboard/) plugin pages, if you open dev tools.

**Extra info**
- I can say with certainty that the Toolbar plugin is part of the problem, because when I commented it out, the problem went away.
- The behaviour exhibited in Firefox is also different from the behaviour exhibited in Chrome (in Firefox, the style is applied to the `pre` element instead of the `div`, but they all get `style: 0px;` instead of a mix), so that's another thing to look out for.

I don't know if this is a problem specific to Prism's website, and/or if there are any other elements at play. I'll poke around more and see what else I can find. Additional help will be appreciated!

Originally discovered here: https://github.com/PrismJS/prism/pull/3104#issuecomment-932996123

**Comment by hoonweiting** [comment]
A little more info, turns out that a `style="margin: 0px;"` is being applied to the `pre` element in Chrome as well, so the only difference between Chrome and Firefox is that Firefox doesn't have the `style` being applied to the code-toolbar `div`.

![image](https://user-images.githubusercontent.com/59229084/136271304-7abace25-df51-4321-a07f-96484982ced8.png)

[My own page](https://hoonweiting.github.io/prismjs-playground/) doesn't have the problem, so I'm guessing the problem is specific to Prism's website. And that led me to poke around other parts of Prism's website.

I commented out the `assets/code.js` script and the problem went away:

![image](https://user-images.githubusercontent.com/59229084/136272802-b2d94362-c6b0-4aba-81f9-a3e266e986d4.png)

I un-commented the script, but commented out the following two lines of code in `assets/code.js` and got the same result (screenshots from Chrome, Firefox respectively):

https://github.com/PrismJS/prism/blob/d38592c5bb1582a2fddefd81c9ad98e403b4ce9f/assets/code.js#L263-L265

![image](https://user-images.githubusercontent.com/59229084/136273629-dc010c19-2d81-437e-b803-0aee56b24fe6.png)

![image](https://user-images.githubusercontent.com/59229084/136275190-c49df610-d969-41d9-b862-b013368c0e56.png)

So at least it isn't a problem with the Toolbar plugin itself, so I guess it's a little less pressing of an issue. But I'm not very sure how to solve it given that it was implemented because of another bug (#1966)?

I'm still not sure what is causing the margin to be transferred inconsistently in Chrome, though. May not be super necessary to figure out why, but I hope it's not a sign of another bug.

**Comment by hoonweiting** [comment]
Okay I *think* the problem is with `requestAnimationFrame();`:
https://github.com/PrismJS/prism/blob/3f24dc727b3e186b6ad6ef651812a15f8cd8fc04/assets/code.js#L255-L267

If my understanding isn't wrong (I'm really basing this off the console logs in Firefox, I've not looked at Autoloader's code), when Autoloader loads a new language, the relevant code blocks are repainted. I think it's easier to explain in point form:
1. Page is initially painted.
2. A bunch of resources load, including most of the languages.
3. `requestAnimationFrame()` runs, and the margins in the `pre` element of all code blocks (which are `.5em 0` for all of the official themes, equivalent to `8px 0`) are transferred to the code-toolbar `div`. Then, the margins of the `pre` elements are set to `0`.
4. The remaining languages load.
5. `requestAnimationFrame()` is run again for the repainted code blocks, and the margins in the `pre` element of these code blocks (which are now `0`, after Step 3) are transferred to the code-toolbar `div`. Then, the margins of the `pre` elements are set to `0` again.

---

So here I can offer a solution, which is to add an if statement to check if the margin transfer had already occurred. If not, then we transfer the margin. (This is on top of modifying the code to use `marginTop` etc instead of `margin`, so that it works in Firefox (and Safari?) as well.)

However, I think this is a bit of an XY problem. The original goal of the margin transfer is to give the toolbar some headspace in code blocks on Prism's site (https://github.com/PrismJS/prism/pull/1966#discussion_r302588858). At first I wanted to suggest adding a rule for `.toolbar` (not `.code-toolbar`) in `assets/styles.css` to increase the value of `top` from the plugin's default (which is `.3em`), while removing the margin transfer entirely. But I wanted to see what the original bug is, so I commented out the fixes in #1966. And it looks...alright. It's got enough headspace. Or maybe I'm missing something. (Screenshot in Firefox, with the `pre` element selected in dev tools. Top edge of the code-toolbar `div` is where the top of the purple is.)

![image](https://user-images.githubusercontent.com/59229084/136446396-29f7ef08-3f2f-4ef7-8722-49c167379717.png)

Anyway, I still went ahead and tried commenting out the margin transfer, and adding `div.code-toolbar div.toolbar { top: .8em; }` to `assets/styles.css`. An extra rule will be needed to reduce the space between two adjacent code blocks (to mimic adjacent margin collapsing), but yeah the result will be similar as the screenshot above.

As a bonus, the latter two solutions above will get rid of the scrollbar for single-line code blocks when the Coy theme is used (which you can see on the [Copy to Clipboard page](https://prismjs.com/plugins/copy-to-clipboard/). I don't know why, but somehow it does.

To summarise, I have three solutions in mind:
1. Ensure margin is transferred only once. (I have not tried this myself, but I think it's possible.)
2. Revert #1966. (I think I'm missing something here honestly,)
3. Remove the margin transfer, and add some CSS rules in `assets/styles.css` to reposition the toolbar.

---

@RunDevelopment what do you think? (And sorry for the text wall!)

**Comment by RunDevelopment** [comment]
> However, Coy doesn't want to play nicely with that

It never does...

Let's just nuke the shadows:

```css
div.code-toolbar > pre:after,
div.code-toolbar > pre:before {
	display: none !important;
}
```

Also, we can see that the usually grey border around code blocks disappeared (even before removing the shadows). This is because the `pre` element doesn't account for the 1px box-shadow of the `code`element. I would argue that this is an oversight with Coy's design and should be fixed by adding a `padding: 1px` to the `pre`. I'll make a PR for that.

> Maybe we can use that method when Coy gets replaced with Coy without Shadows in future?

Yes, we should be. Coy is implemented in a pretty weird way which causes loads of problems but enables these cool shadows. Coy without Shadows takes the standard approach to themes but can't implement the shadows (easily).

> But I'm wondering why the `overflow: auto;` property is needed at all?

Because that "cool" theme selector on the right side messes with the layout of pages and overlaps with elements. `overflow: auto` fixes that. Here's the `div` with `overflow: auto`:

![image](https://user-images.githubusercontent.com/20878432/136547010-57ef1c0c-3296-43ca-b213-5d332a01b4c7.png)

![image](https://user-images.githubusercontent.com/20878432/136547173-9821156a-72b2-4605-b536-a646c0df3c1b.png)


and without:

![image](https://user-images.githubusercontent.com/20878432/136547096-192d39cb-33c7-4c15-99a8-0142bab00b29.png)

![image](https://user-images.githubusercontent.com/20878432/136547142-a20714e8-2af7-4c96-af6f-fc4620579145.png)

**Comment by hoonweiting** [comment]
:") I must say Coy has really pretty colours (and the shadows are a really nice touch), it's just such a shame that its shadows don't play nicely with the rest of Prism!

Anyway to confirm, what's the plan? I am a little lost :sweat_smile: (I added the 1px padding on my end too for the screenshots)

| Here's margin transfer with nuked shadows: | And here's shifted toolbar with negative `div.code-toolbar:not(:first-of-type)` margins: |
| ----- | ----- |
| ![image](https://user-images.githubusercontent.com/59229084/136570693-40caa519-e18a-4954-aa0e-47abc54a3d58.png) | ![image](https://user-images.githubusercontent.com/59229084/136570786-b79b58cc-2ca1-4d95-a112-5dd4e3ddafe2.png) |
| We probably have to nudge the toolbar's position just a touch higher to account for single-line code blocks in Coy? | The shadows don't appear in the darker-coloured rows... I'm soooooo done |

(Then again, sounds like the shadow nuking is a joke HAHA whoopsie)

> Also, we can see that the usually grey border around code blocks disappeared (even before removing the shadows).

Whoa, nice catch!

> Because that "cool" theme selector on the right side messes with the layout of pages and overlaps with elements.

Ah that's what I missed \*facepalm\* thanks!

**Comment by hoonweiting** [comment]
Ah I figured out why the shadows don't show up on the darker-coloured rows. It's because the z-index of the shadows are less than 0, and setting a z-index of `1` on the `pre` will fix it.

**Comment by hoonweiting** [comment]
Hmm, I guess I am a bit concerned about removing the shadows on Prism's site, because it's the only way users can 'preview' the theme. If they were to download it, they may be confused as to why their downloaded version has shadows, whereas there aren't shadows on Prism's site.

---

Also here's a screenshot of my previous comment, with the additional z-index properties set (sorry for not including it earlier):
![image](https://user-images.githubusercontent.com/59229084/136646206-3db7ef26-9422-4f17-8f89-eb883b5042f6.png)
